src/mongoc/mongoc-uri.c

```diff
@@ -704,6 +704,54 @@ mongoc_uri_get_mechanism_properties (const mongoc_uri_t *uri, bson_t *properties
 
 
 static void
+_mongoc_uri_build_read_prefs (mongoc_uri_t *uri) /* IN */
+{
+   mongoc_read_prefs_t *read_prefs;
+   const char *str;
+   bson_iter_t iter;
+
+   BSON_ASSERT(uri);
+
+   read_prefs = mongoc_read_prefs_new(MONGOC_READ_PRIMARY);
+
+   if (bson_iter_init_find_case(&iter, &uri->options, "slaveok") &&
+       BSON_ITER_HOLDS_BOOL(&iter) &&
+       bson_iter_bool(&iter)) {
+      mongoc_read_prefs_set_mode(read_prefs, MONGOC_READ_SECONDARY_PREFERRED);
+   }
+
+   if (bson_iter_init_find_case(&iter, &uri->options, "readpreference") &&
+       BSON_ITER_HOLDS_UTF8(&iter)) {
+      str = bson_iter_utf8(&iter, NULL);
+
+      if (0 == strcasecmp("primary", str)) {
+         mongoc_read_prefs_set_mode(read_prefs, MONGOC_READ_PRIMARY);
+      } else if (0 == strcasecmp("primarypreferred", str)) {
+         mongoc_read_prefs_set_mode(read_prefs, MONGOC_READ_PRIMARY_PREFERRED);
+      } else if (0 == strcasecmp("secondary", str)) {
+         mongoc_read_prefs_set_mode(read_prefs, MONGOC_READ_SECONDARY);
+      } else if (0 == strcasecmp("secondarypreferred", str)) {
+         mongoc_read_prefs_set_mode(read_prefs, MONGOC_READ_SECONDARY_PREFERRED);
+      } else if (0 == strcasecmp("nearest", str)) {
+         mongoc_read_prefs_set_mode(read_prefs, MONGOC_READ_NEAREST);
+      } else {
+         MONGOC_WARNING("Unsupported readPreference value [readPreference=%s].", str);
+      }
+   }
+
+   if (!bson_empty(&uri->read_prefs_tags)) {
+      /* Warn on conflict, since read preference will be validated later */
+      if (mongoc_read_prefs_get_mode(read_prefs) == MONGOC_READ_PRIMARY) {
+         MONGOC_WARNING("Primary read preference mode conflicts with tags.");
+      }
+      mongoc_read_prefs_set_tags(read_prefs, &uri->read_prefs_tags);
```

Comment:
Since this function is always called between options parsing and the URI constructor returning, we could immediately destroy `&uri->read_prefs_tag` here and modify the should-be-deprecated `mongoc_uri_get_read_prefs()` function to fetch tags directly from `uri->read_prefs`.